File: reverse-engineering/ghidra_scripts/decompile_all.py
# Ghidra headless post-script (Jython 2.7 / Ghidra 11-12)
# Decompile every function and dump readable C to $GHIDRA_OUT.
# Optional $GHIDRA_FILTER: only emit functions whose name contains this substring.

from ghidra.app.decompiler import DecompInterface
from ghidra.util.task import ConsoleTaskMonitor
import os

out_path = os.environ.get("GHIDRA_OUT", "decompiled.c")
filt = os.environ.get("GHIDRA_FILTER", "")

prog = currentProgram
ifc = DecompInterface()
ifc.openProgram(prog)
mon = ConsoleTaskMonitor()

fm = prog.getFunctionManager()
funcs = fm.getFunctions(True)

out = open(out_path, "w")
count = 0
total = 0
for f in funcs:
    total += 1
    name = f.getName()
    if filt and filt not in name:
        continue
    try:
        res = ifc.decompileFunction(f, 30, mon)
    except Exception as e:
        continue
    if res is None or not res.decompileCompleted():
        continue
    code = res.getDecompiledFunction().getC()
    out.write("// ===== %s @ %s =====\n" % (name, f.getEntryPoint()))
    out.write(code)
    out.write("\n\n")
    count += 1

out.close()
printf = None
print("GHIDRA: dumped %d / %d functions to %s" % (count, total, out_path))

File: reverse-engineering/ghidra_scripts/DecompileAll.java
// Ghidra headless post-script: decompile every function to $GHIDRA_OUT.
// Optional $GHIDRA_FILTER: only emit functions whose name contains the substring.
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.DecompInterface;
import ghidra.app.decompiler.DecompileResults;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.PrintWriter;

public class DecompileAll extends GhidraScript {
    @Override
    public void run() throws Exception {
        String outPath = System.getenv("GHIDRA_OUT");
        if (outPath == null || outPath.isEmpty()) outPath = "decompiled_all.c";
        String filter = System.getenv("GHIDRA_FILTER");
        if (filter == null) filter = "";

        DecompInterface ifc = new DecompInterface();
        ifc.openProgram(currentProgram);

        PrintWriter w = new PrintWriter(new BufferedWriter(new FileWriter(new File(outPath))));
        FunctionManager fm = currentProgram.getFunctionManager();
        int count = 0, total = 0;
        for (Function f : fm.getFunctions(true)) {
            total++;
            String name = f.getName();
            if (!filter.isEmpty() && !name.contains(filter)) continue;
            try {
                DecompileResults res = ifc.decompileFunction(f, 30, monitor);
                if (res != null && res.decompileCompleted()) {
                    w.println("// ===== " + name + " @ " + f.getEntryPoint() + " =====");
                    w.println(res.getDecompiledFunction().getC());
                    w.println();
                    count++;
                }
            } catch (Exception e) {
                // skip unliftable functions (Enigma VM stubs, bad data)
            }
        }
        w.close();
        println("GHIDRA: dumped " + count + " / " + total + " functions to " + outPath);
    }
}
